### PR TITLE
Add support for caching temporary session credentials

### PR DESCRIPTION
**Pull request by mixja**
This PR adds support for caching temporary session credentials, using the same mechanism as used in the AWS CLI.

The main benefit of this feature is a much improved experience for users that require MFA for API access.  Prior to this PR, the experience is that users have to re-enter MFA token each time they run the command:

```
$ awslogs groups
Enter MFA code: ******
/my/log/group

$ awslogs get /my/log/group
Enter MFA code: ******
...
...
```

With this PR, the user experience will be that they only need to enter an MFA code once for the lifetime of the session credentials (e.g. once per hour):

```
$ awslogs groups
Enter MFA code: ******
/my/log/group

$ awslogs get /my/log/group
...
...
```

Caching is invoked only if the `assume-role` provider is active, which happens when you are using a profile that assumes a role:

```
$ export AWS_PROFILE=some-account-admin
$ cat ~/.aws/config
[profile some-account-admin]
source_profile=users-account
role_arn=arn:aws:iam::012345678912:role/admin
mfa_serial=arn:aws:iam::219876543210:mfa/justin.menga
region=us-west-2
```

### CODE REVIEW

**Comment by ryanlitalien** [comment]
@jorgebastida Can we get this merged in and published? It'd be really helpful for my day to day.

**Comment by jorgebastida** [comment]
Well, after reading this PR I went a bit on a rabbit hole. 

tl;dr I don't thinks ``awslogs`` is the place where 2fa authentication should be handled, so I wrote a tool I wanted to write for a while now. You can find it here https://github.com/jorgebastida/aws2fa

Once you use it, you should be able to use awslogs normally

**Comment by mixja** [comment]
@jorgebastida - I don't think you quite understand - this PR does not implement MFA, it simply adds support for reading and writing to the AWS CLI generated session cache.  

Also the AWS CLI natively supports MFA so you don't need another utility for this (i.e. https://github.com/jorgebastida/aws2fa) - see http://docs.aws.amazon.com/cli/latest/userguide/cli-roles.html#cli-roles-mfa

If you try and use `awslogs` currently with an MFA enabled CLI profile you will be prompted for MFA token (this is part of the boto3 library).  If you run `awslogs` again you will be prompted for MFA again ...and again... and again - without support for the session cache, you will always have this problem.

**Comment by jorgebastida** [comment]
I understand. My view is that I think is not practical that every app deals with this credentials cache by themselves. Having a number of cached sessions credentials all around my computer for every single app feels odd (or at least not aligned with my workflow).

When I get into my computer I want to 2fa my keys for a number of minutes I feel confortable with, and then I expect all apps to use those credentials. Once the time expires, all apps loose access to my account.

**Comment by mixja** [comment]
@jorgebastida - I agree it seems impractical to deal with the credentials cache, but until boto3 includes such functionality natively, this approach makes life a lot easier for the user.  Note that the credentials exist per _AWS profile_ not per application.  

As I understand it I'm guessing your workflow is to inject temporary session credentials into your local environment and then run awslogs.  This requires a number of manual steps or custom scripts/utilities (e.g. your aws2fa).

The experience of leveraging the session cache:

```
$ awslogs <cmd> --profile my-mfa-profile
Enter MFA Token: ******
...
...
$ awslogs <cmd> --profile my-mfa-profile
...
...
```

or

```
$ export AWS_PROFILE=my-mfa-profile
$ awslogs <cmd>
Enter MFA Token: ******
...
...
$ awslogs <cmd>
...
...
```

The alternative of Injecting credentials into environment (best case scenario, worst case you have multiple steps before running `awslogs`:

```
$ run_a_custom_script_or_program.blah
Enter MFA Token: ******
$ awslogs <cmd>
...
...
$ awslogs <cmd>
...
...
```

I think by implementing support for the session cache, you make it simpler for the end user (no custom utilities or scripts to deal with).

**Comment by mixja** [comment]
@jorgebastida - botocore now includes the credential cache that was implemented in this PR.

Are you willing to add this feature in given it is part of botocore now?  Note that the lifetime of credentials in the credential cache is up to **one hour**, so this provides a real convenience if you have MFA and need to run awslogs multiple times within one hour.

If you are willing to include this feature I will code it up and write the tests, but don't want to waste my time if you are not interested?

Thanks

**Comment by mixja** [comment]
Looks like support for MFA was finally added in here - https://github.com/jorgebastida/awslogs/pull/222 - much better experience now using awslogs :)
